nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/truncate_path_local_action.hpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Francisco Martin Rico
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__PLUGINS__ACTION__TRUNCATE_PATH_LOCAL_ACTION_HPP_
+#define NAV2_BEHAVIOR_TREE__PLUGINS__ACTION__TRUNCATE_PATH_LOCAL_ACTION_HPP_
+
+#include <memory>
+#include <string>
+
+#include "nav_msgs/msg/path.hpp"
+
+#include "behaviortree_cpp_v3/action_node.h"
+#include "tf2_ros/buffer.h"
+
+namespace nav2_behavior_tree
+{
+
+/**
+ * @brief A BT::ActionNodeBase to shorten path by some distance
+ */
+class TruncatePathLocal : public BT::ActionNodeBase
+{
+public:
+  /**
+   * @brief A nav2_behavior_tree::TruncatePathLocal constructor
+   * @param xml_tag_name Name for the XML tag for this node
+   * @param conf BT node configuration
+   */
+  TruncatePathLocal(
+    const std::string & xml_tag_name,
+    const BT::NodeConfiguration & conf);
+
+  /**
+   * @brief Creates list of BT ports
+   * @return BT::PortsList Containing basic ports along with node-specific ports
+   */
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<nav_msgs::msg::Path>("input_path", "Original Path"),
+      BT::OutputPort<nav_msgs::msg::Path>(
+        "output_path", "Path truncated to a certain distance around robot"),
+      BT::InputPort<double>(
+        "distance_forward", 8.0,
+        "Distance in forward direction"),
+      BT::InputPort<double>(
+        "distance_backward", 4.0,
+        "Distance in backward direction"),
+      BT::InputPort<std::string>(
+        "robot_frame", "base_link",
+        "Robot base frame id"),
+      BT::InputPort<double>(
+        "transform_tolerance", 0.2,
+        "Transform lookup tolerance"),
+      BT::InputPort<geometry_msgs::msg::PoseStamped>(
+        "pose", "Manually specified pose to be used"
+        "if overriding current robot pose"),
+      BT::InputPort<double>(
+        "angular_distance_weight", 0.2,
```

Comment:
This attitude could be quite helpful, but it would require to use the path pruning which is performed by the controllers, i.e. assigning whole path to a field when a goal is received and then trimming it step-by-step. As opposed to controller server/plugins, this node is not an action server, it's not even stateful, it simply gets the current path from the blackboard every time and returns a trimmed portion of it.

Using `angular_distance_weight` has solved the loop problem for us, making it possible to keep the node that simple. However, if required, it is possible to mimic the stateful behavior of RPP/DWB. I'd prefer the behavior as an option though, since non-statefulness has its advantages too (no risk of trimming off something that should not be trimmed in case of a temporal anomaly in the inputs, without a chance to return it back), i.e. to use the current non-stateful behavior if `max_robot_pose_search_dist` is `.inf`.

So if decided to implement the path pruning and `max_robot_pose_search_dist`, my idea would be to keep track of
 * `last_robot_pose_index_` to be used as a `begin` for computing the `closest_pose_upper_bound` (this would mimic the path pruning)
 * keep track of `last_path_` to compare it against current `path`. If the path has changed, reset `last_robot_pose_index_` to 0 (this mimics the [`setPlan()`](https://github.com/ros-planning/navigation2/blob/4234a6ea27de2094dde98a5f928522d72b139227/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L631) method of RPP/DWB which basically resets all trimming)

It's not a 100% elegant solution, so it's worth considering, but should be quite easy to implement.

Reply:
I'm not sure I understand, you're processing the path here anyway, if you grab the path from the blackboard and create a local variable with the path value, you can do whatever you want. You have a `path` and a `robot_pose`, this is no different inputs than those have. I'm not suggesting you override anything, I'm suggesting you remove the `angular_distance_weight` and the custom `poseDistance` that uses it and instead use the method https://github.com/ros-planning/navigation2/blob/main/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L670-L682 to find the closest point on the path, since that style is mirrored in RPP, DWB, and I believe in 1 other place to do that same operation.

Then once you have the point in the path closest to the robot, you can use the integrated distance `nav2_util` (maybe needs to be modified for inputs or create a new instance of it) to find the segment backward and forward of interest based on your parameters, rather than implementing those inline in the BT node. I think integrating N distance backward or forward from a point on the path could be a useful general utility for other applications, don't you think?

Reply:
Yeah, it's a great new util for this purpose, I'm already using it in my code, just haven't pushed that yesterday, since this discussion hasn't been resolved yet... No need to add a special "reverse" case, it works well with `std::reverse_iterator`s